### PR TITLE
fix: V1 trunk proofs + count==0 hash verification bypass

### PR DESCRIPTION
**Pull request by QuantumExplorer**
## Summary

- Adds V1 trunk proof generation using `LayerProof`/`ProofBytes::Merk` (previously only V0 `MerkOnlyLayerProof` was supported)
- Fixes a **high-severity security vulnerability** in trunk proof verification where an attacker can forge an empty-tree result for a non-empty CountTree

### Vulnerability (V0)

In `verify_trunk_chunk_proof_v0`, when `count == 0` the function returns early **without** running the `combine_hash(value_hash(value_bytes), lower_hash)` verification loop. Since merk `KVValueHash` nodes store `value_hash` separately from value bytes, an attacker can:

1. Take a valid trunk proof for a non-empty CountSumTree (100 items)
2. Replace the serialized element bytes to set `count = 0`
3. Keep the original `value_hash` intact in the `KVValueHash` node
4. The merk proof still verifies (it uses the embedded `value_hash`)
5. The `count == 0` fast-path returns an empty result, skipping hash verification

### Fix (V1)

The V1 verifier **always** runs the `combine_hash` verification chain, using `NULL_HASH` as the lower hash for empty trees. Forged element bytes produce a different `value_hash`, so `combine_hash(H(forged_bytes), NULL_HASH) != expected_hash`, and the proof is rejected.

## Test plan

- [x] `cargo test -p grovedb --lib trunk_proof_tests` — all 7 tests pass
- [x] `cargo test -p grovedb --lib chunk_branch_proof` — all 3 tests pass
- [x] `cargo clippy -p grovedb -- -D warnings` — clean
- [x] Security regression test (`test_trunk_proof_v1_rejects_forged_count_zero`) proves V1 rejects and V0 accepts the attack
- [ ] CI passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Enabled trunk-chunk proofs in GroveDB v3 and added a V1 proof format with Merkle-style bytes and multi-layer proof support.

* **Bug Fixes**
  * Strengthened bottom-up hash-chain verification across layers and tightened handling of empty and edge-case trees.

* **Tests**
  * Expanded coverage to validate the V1 proof format, tampering scenarios, multi-level integrity, and cross-version behaviors.

### CODE REVIEW

File: grovedb/src/tests/trunk_proof_tests.rs [hunk 1]
```diff
@@ -522,4 +588,969 @@ mod tests {
...
+            bincode::encode_to_vec(&GroveDBProof::V1(proof_v1), config).expect("encode");
+
+        let result = GroveDb::verify_trunk_chunk_proof(&proof_bytes, &empty_query, grove_version);
+        assert!(result.is_err(), "should reject empty path");
+        let err_msg = format!("{:?}", result.unwrap_err());
+        assert!(
+            err_msg.contains("not a count tree") || err_msg.contains("empty path"),
+            "wrong error: {}",
+            err_msg
+        );
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply grovedb/src/tests/trunk_proof_tests.rs:1379]: This is Claude. Already addressed in 44fd14bd — moved the empty-path guard to the top of `verify_trunk_chunk_proof_v1` (before the for-loop), and tightened the assertion to require specifically `"empty path"`.

File: grovedb/src/operations/proof/verify.rs [hunk 2]
```diff
@@ -2190,6 +2190,262 @@ impl GroveDb {
...
+                    return Err(Error::InvalidProof(
+                        PathQuery::new_unsized(current_path.clone(), key_query),
+                        "V1 trunk proof: expected Merk proof bytes at path layer".to_string(),
+                    ));
+                }
+            };
+
+            let (layer_root_hash, result) = key_query
+                .execute_proof(merk_proof_bytes, None, true, PROOF_VERSION_LATEST)
+                .unwrap()
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply grovedb/src/operations/proof/verify.rs:2249]: This is Claude. Not fixing — this is a false positive.

`PROOF_VERSION_LATEST` is a compile-time `const u16 = 1` defined in `merk/src/proofs/query/verify.rs:21`, which is part of the same Cargo workspace. It's not an external dependency that can change independently. Any version bump is a deliberate workspace-wide decision that affects all code paths simultaneously.

The same constant is used identically in `verify_layer_proof_v1` (line 461) for regular query proofs. Pinning only the trunk proof path would be inconsistent with the rest of the V1 verification code. When a new merk proof version is introduced, it will require a new GroveDB proof version anyway (V2), at which point both paths would be updated together.

File: grovedb/src/tests/trunk_proof_tests.rs [hunk 3]
```diff
@@ -522,4 +588,969 @@ mod tests {
...
+        let config = bincode::config::standard()
+            .with_big_endian()
+            .with_no_limit();
+        let tampered = bincode::encode_to_vec(&GroveDBProof::V1(proof_v1), config).expect("encode");
+
+        let result = GroveDb::verify_trunk_chunk_proof(&tampered, &query, grove_version);
+        assert!(
+            result.is_err(),
+            "should reject proof with invalid op sequence"
+        );
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply grovedb/src/tests/trunk_proof_tests.rs:1411]: This is Claude. Fixed in 11230dc3 — the test now asserts `err_msg.contains("Failed to execute V1 trunk proof")` to anchor to the specific execution-failure branch.

File: grovedb/src/operations/proof/verify.rs [hunk 4]
```diff
@@ -2190,6 +2190,262 @@ impl GroveDb {
...
+        let mut elements = BTreeMap::new();
+        let mut leaf_keys = BTreeMap::new();
+        Self::extract_elements_and_leaf_keys(
+            &target_tree,
+            &mut elements,
+            &mut leaf_keys,
+            0,
+            max_depth,
+            grove_version,
+        )?;
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply grovedb/src/operations/proof/verify.rs:2445]: This is Claude. Not fixing in this PR — this is a pre-existing issue, not introduced by these changes.

The `execute()` function (used for the target chunk in both V0 and V1 trunk verification) accepts `KVValueHash` nodes and uses the embedded `value_hash` directly without recomputing from `value_bytes` (see `merk/src/proofs/tree.rs:134-138`, which explicitly comments "cannot verify hash(value) == value_hash"). This is by design for subtree nodes where `value_hash` is a `combine_hash(element_hash, subtree_root_hash)`.

For the target chunk (leaf-level elements), an attacker could theoretically substitute `KV(key, value)` with `KVValueHash(key, forged_value, real_hash)`. However:

1. This affects V0 trunk verification identically (same `execute()` call) — it's not a regression
2. The prover-side constraint (test at lines 202-249) verifies only `KV`/`Hash` nodes are generated
3. The proper fix belongs at the merk execution layer or as a separate hardening PR, not in this trunk-proof-specific change

[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply grovedb/src/operations/proof/verify.rs:2445]: This is Claude. Update: this has now been fixed in commit a07568e3.

`extract_elements_and_leaf_keys` now validates embedded `value_hash` for all `KVValueHash*` node variants:

- **`KVValueHash` / `KVValueHashFeatureType`**: verifies `H(value) == value_hash`
- **`KVValueHashFeatureTypeWithChildHash`**: verifies `combine_hash(H(value), child_hash) == value_hash` (was already present)
- **`KVRefValueHash` / `KVRefValueHashCount`**: rejected outright (should never appear in trunk/branch proofs)

Two attack tests added: one forging `KVValueHash` and one forging `KVValueHashFeatureType`, both asserting "value hash mismatch" rejection.
